scripts/iam-tools/aws_creds_check.sh

```diff
@@ -0,0 +1,115 @@
+#!/bin/bash
+
+# Note: because we source this in the login shell, it needs to be compatible with bash and zsh.
+# There are several places where -n "$ZSH_NAME" is tested to see if we're running zsh, and then
+# the zsh version of the syntax is used instead of the bash version.
+
+# only bash v4+ and zsh are supported. do nothing for other shells
+if [ -n "$BASH_VERSION" -a "$BASH_VERSINFO" -gt 3 -o "$ZSH_NAME" ]; then
+
+function _check_creds()
+{
+    local -A accounts
+    local account_count creds_actual name num line regex now credfile=~/.aws/credentials expfile=~/.aws/credentials_expiration
+
+    if ! private_encfs_mounted; then
+        echo "~/.private not mounted. Skipping AWS credential check"
+        return
+    fi
+
+    # if ~/.private/.aws doesn't exist, create it
+    test -d ~/.private/.aws || mkdir ~/.private/.aws
+    # if ~/.private/.aws/credentials doesn't exist, create it
+    test -f ~/.private/.aws/credentials || { touch ~/.private/.aws/credentials; chmod 600 ~/.private/.aws/credentials; }
+    # if ~/.aws doesn't exist, create it as a symlink to ~/.private/.aws
+    test -d ~/.aws || ln -s ~/.private/.aws ~/.aws
+    # if ~/.aws is a directory and ~/.aws/credentials doesn't exist, create it as a symlink to ~/.private/.aws/credentials
+    test -d ~/.aws -a ! -h ~/.aws -a ! -f .aws/credentials && ln -s ~/.private/.aws/credentials ~/.aws/credentials
+    creds_actual=$(readlink -f ~/.aws/credentials)
+    if [ "$creds_actual" != ~/.private/.aws/credentials ]; then
+        echo "WARNING WARNING WARNING WARNING WARNING WARNING WARNING"
+        echo ~/.aws/credentials must be stored in the encrypted directory ~/.private
+        echo "WARNING WARNING WARNING WARNING WARNING WARNING WARNING"
+    fi
+
+    [ -f "$expfile" ] && read -r line < "$expfile"
+    line=$((line+0))
+    if [ -z "$ZSH_NAME" ]; then
+        #    like 'date +%s' without a fork/exec
+        now="$(printf '%(%s)T' -1)"
+    else
+        # zsh's printf doesn't have time formatters so we're forced to fork/exec date
+        now="$(date +%s)"
+    fi
+    if [ $((now + 60 * 60 * 24 * 7)) -gt "$line" ]; then
```

Comment:
Because you're doing the same date math several times in a row, it'd make this clearer to abstract those out into variables.

`HOUR=60*60`
`DAY=${HOUR}*24`
`WEEK=${DAY}*7`
`if [ $((now+${WEEK})) -gt "$line" ]; then`